pulsar-broker/src/main/java/org/apache/pulsar/compaction/TwoPhaseCompactor.java

```diff
@@ -0,0 +1,282 @@
+
+    }
+
+    private void scheduleTimeout(CompletableFuture<RawMessage> future) {
+        Future<?> timeout = scheduler.schedule(() -> {
+                future.completeExceptionally(new TimeoutException("Timeout"));
+            }, 10, TimeUnit.SECONDS);
+        future.whenComplete((res, exception) -> {
+                timeout.cancel(true);
+            });
+    }
+
+    private CompletableFuture<Long> phaseTwo(RawReader reader, MessageId from, MessageId to,
+                                             Map<String,MessageId> latestForKey, BookKeeper bk) {
+        return createLedger(bk).thenCompose(
+                (ledger) -> phaseTwoSeekThenLoop(reader, from, to, latestForKey, bk, ledger));
+    }
+
+    private CompletableFuture<Long> phaseTwoSeekThenLoop(RawReader reader, MessageId from, MessageId to,
+                                                         Map<String, MessageId> latestForKey,
+                                                         BookKeeper bk, LedgerHandle ledger) {
+        CompletableFuture<Long> promise = new CompletableFuture<>();
+
+        reader.seekAsync(from).thenCompose((v) -> {
+                Semaphore outstanding = new Semaphore(MAX_OUTSTANDING);
+                CompletableFuture<Void> loopPromise = new CompletableFuture<Void>();
+                phaseTwoLoop(reader, to, latestForKey, ledger, outstanding, loopPromise);
+                return loopPromise;
+            }).thenCompose((v) -> closeLedger(ledger))
+            .thenCompose((v) -> reader.acknowledgeCumulativeAsync(
+                                 to, ImmutableMap.of(COMPACTED_TOPIC_LEDGER_PROPERTY, ledger.getId())))
+            .whenComplete((res, exception) -> {
+                    if (exception != null) {
+                        deleteLedger(bk, ledger)
+                            .whenComplete((res2, exception2) -> {
+                                    // complete with original exception
+                                    promise.completeExceptionally(exception);
+                                });
+                    } else {
+                        promise.complete(ledger.getId());
+                    }
+                });
+        return promise;
+    }
+
+    private void phaseTwoLoop(RawReader reader, MessageId to, Map<String, MessageId> latestForKey,
+                              LedgerHandle lh, Semaphore outstanding, CompletableFuture<Void> promise) {
+        reader.readNextAsync().whenComplete(
+                (m, exception) -> {
+                    try {
+                        if (exception != null) {
+                            promise.completeExceptionally(exception);
+                            return;
+                        } else if (promise.isDone()) {
+                            return;
+                        }
+                        MessageId id = m.getMessageId();
+                        String key = extractKey(m);
+
+                        if (latestForKey.get(key).equals(id)) {
```

Comment:
what if by the time phaseTwo runs, expected message with latestKey has been deleted and reader will never be able to read?

Reply:
The fact that we have a subscription (in the RawReader) should keep it there. I'll added a task to add a test to ensure this.